Interactivity API: Refactor internal proxy and signals system

DAreRodz opened the pull request:
## What?

The main goal of this refactoring is to have more control when defining state/context properties. The idea is to keep a unique [computed](https://github.com/preactjs/signals?tab=readme-ov-file#computedfn) for each of these properties *per scope*, which tracks changes even when its value is replaced with a getter. This way, properties can contain an initial value and be replaced with a getter later on or vice-versa, ensuring that directives update accordingly in this case. This is especially relevant for stores that are loaded asynchronously, like the `core/router` store from the `@wordpress/interactivity-router` module.

Also, we can take this opportunity to remove an external dependency ([deepsignal](https://github.com/luisherranz/deepsignal/)), keeping the code specific to the Interactivity API and making it more maintainable.

### Tasks

- [x]  Outline code structure
- [x]  Add tests to the state proxy handlers
- [x]  Ensure computeds work without scope
- [x]  Ensure objects maintain references between different namespaces
- [x] Add tests to the store proxy handlers
- [x] Add tests to `peek()`
- [x]  Review naming
- [x]  Review types and add TSDocs

### Next
I think these tasks should be addressed in a subsequent PR:
-  Use maps to store the context objects by namespace
-  Refactor context and related tests

## Testing Instructions

No testing instructions yet.

DAreRodz committed:
Add computation tests

DAreRodz committed:
WIP

DAreRodz committed:
Refactor code a little

DAreRodz committed:
Add tests for getters with scope

DAreRodz committed:
Fix getter call

DAreRodz committed:
Fix a problem with object references

DAreRodz committed:
Restore previous packages

DAreRodz committed:
Fix current tests

DAreRodz committed:
Add store handlers

DAreRodz committed:
Move namespaces to properties

DAreRodz committed:
Fix for..in on new properties

DAreRodz committed:
More WIP

DAreRodz committed:
A bit of refactoring

DAreRodz committed:
Move withScope inside PropSignal logic

DAreRodz committed:
Add TODO comment inside `peek()`

DAreRodz committed:
Attempt to make the context work

DAreRodz committed:
Add signals-core dependency

DAreRodz committed:
Reorganize code

DAreRodz committed:
Create proxies folder

DAreRodz committed:
Fix lint error

DAreRodz committed:
Rename DEFAULT_SCOPE to NO_SCOPE

DAreRodz committed:
Move store initialization inside init

DAreRodz committed:
Return the computed value with `peek()`

DAreRodz committed:
Fix store root assignments

DAreRodz committed:
Remove deepsignal

DAreRodz committed:
Make peek() return only the value inside signalValue

DAreRodz committed:
Fix context proxification

DAreRodz committed:
Handle functions inside state

DAreRodz committed:
Rename Property to PropSignal

DAreRodz committed:
Fix withScope in this PR

DAreRodz committed:
Move store root logic to store proxy handlers

DAreRodz committed:
Enable skipped test for non-initialized getters

DAreRodz committed:
Rename get(State|Store)Proxy to proxify(State|Store)

DAreRodz committed:
Make `getContext` throw when there is no scope

DAreRodz committed:
Fix `proxifyState` types

DAreRodz committed:
Rename some variables in tests

DAreRodz committed:
Add test for object reference keeping

DAreRodz committed:
Rename test suite for state proxy

DAreRodz committed:
Add tests for getters and functions with scope

DAreRodz committed:
Add tests for prop subscription inside functions

DAreRodz committed:
Allow functions to use `this`

DAreRodz committed:
Minor fixes

DAreRodz committed:
Add tests to store proxies

DAreRodz committed:
Throw an error when an object cannot be proxified

DAreRodz committed:
Change peek implementation

DAreRodz committed:
Add tests for peek and unsupported structures

DAreRodz committed:
Test peeking getters that access scope or other namespaces

DAreRodz committed:
Ignore well-known symbols

DAreRodz committed:
Minor comment format fix

DAreRodz committed:
Move namespace arg to first position in proxify functions

michalczaplinski commented on packages/interactivity/src/proxies/registry.ts:
```diff
@@ -0,0 +1,34 @@
+	handlers: ProxyHandler< T >
+): T => {
+	if ( ! shouldProxy( obj ) ) {
+		throw Error( 'This object can be proxified.' );
```
Shouldn't this say `"can't be proxified"`? 🙂

DAreRodz replied:
Of course. 😂

DAreRodz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/62734/commits/1997db5db92672203bfd73e908322679071be0e6

michalczaplinski commented on packages/interactivity/src/proxies/registry.ts:
```diff
@@ -0,0 +1,34 @@
+	if ( typeof val !== 'object' || val === null ) {
+		return false;
+	}
+	return ! ignore.has( val ) && supported.has( val.constructor );
```
Why do we need a separate `ignore` WeakSet here? Can't we re-use `proxyToNS` for this purpose instead and check if `proxyToNS.has(val)` instead?

DAreRodz replied:
Yeah, we could reuse `proxyToNS`. I was considering changing the `ignore` prop, as no other entities but proxies are being "ignored" here.

DAreRodz replied:
Done in https://github.com/WordPress/gutenberg/pull/62734/commits/1997db5db92672203bfd73e908322679071be0e6

michalczaplinski replied:
Thanks! For very large stores that should save us a bit of memory! 🙂

michalczaplinski commented on packages/interactivity/src/directives.tsx:
```diff
@@ -127,21 +130,18 @@ const proxifyContext = ( current: object, inherited: object = {} ): object => {
 /**
  * Recursively update values within a deepSignal object.
  *
- * @param target A deepSignal instance.
- * @param source Object with properties to update in `target`.
+ * @param proxy  A deepSignal instance.
```
We should probably remove references to `deepSignal` in the docstring.

DAreRodz replied:
Sure thing.

michalczaplinski committed:
chore: Update jest.config.js to stop ignoring deepsignal because we removed it as dependency

michalczaplinski commented on packages/interactivity/src/proxies/store.ts:
```diff
@@ -0,0 +1,56 @@
+export const proxifyStore = < T extends object >(
+	namespace: string,
+	obj: T,
+	isNotRoot = false
+): T => {
+	const proxy = createProxy( namespace, obj, storeHandlers );
+	if ( proxy && ! isNotRoot ) {
```
Could we call the parameter `isRoot` and then change the `&& ! isNotRoot` to `&& isRoot` ?

DAreRodz replied:
Actually, that was my first approach. I just changed it because I felt the default case would be creating a root object. I can change the value to be `true` by default, though. 🙂

gziolo replied:
Let's avoid the variable names with negations, as it's hard to reason about `! isNotRoot` in contrast to `isRoot`.

DAreRodz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/62734/commits/5c17dd76651fb88620a4c48b4eb4ad45a6f11f5f.

michalczaplinski commented on packages/interactivity/src/proxies/state.ts:
```diff
@@ -0,0 +1,200 @@
+		 * is a well-known symbol, then it should not be processed.
+		 */
+		if (
+			peeking ||
```
I might be misunderstanding something, but is this not a potential race condition? 

Can it happen that `peeking === true` because some key `A` is being peeked at, but at the very same time, the `get()` handler is invoked for some **different** key `B`?

luisherranz replied:
Those operations are synchronous and, if they are nested, they should all peek. But feel free to add a test to verify that's the case 🙂

DAreRodz replied:
Yup, it works like Luis mentioned here. ☝️

michalczaplinski replied:
Gotcha, thanks guys!

michalczaplinski reviewed (commented):
I left a couple of comments on the technical side of things and pushed a minor commit too.

DAreRodz committed:
Add comments to proxy registry

DAreRodz committed:
Remove namespace from PropSignal

DAreRodz committed:
Remove unused `peekValueSignal` method

DAreRodz committed:
Simplify PropSignal methods

DAreRodz committed:
Add TSDocs to PropSignal

DAreRodz committed:
Disable unused vars lint rule in state-proxy tests

DAreRodz committed:
Remove descriptor alias

DAreRodz committed:
Expand `getProxyNs` docs

luisherranz commented:
@darerodz I have linked [this issue](https://github.com/WordPress/gutenberg/issues/63651) to be closed when this pull request is merged with a test that makes sure that plain arrays cannot be used at the root of the context.

DAreRodz committed:
Add more comments in `state`

DAreRodz committed:
Refactor state functions and add tsdocs

DAreRodz committed:
Fix some grammar issues

DAreRodz committed:
Fix default namespace for setters

DAreRodz committed:
Replace `isNotRoot` with `isRoot`

DAreRodz committed:
Update comments in store.ts

DAreRodz committed:
Move `isPlainObject` to utils

DAreRodz committed:
Remove remaining deepSignal references

DAreRodz commented on packages/interactivity/src/proxies/index.ts:
```diff
@@ -0,0 +1,5 @@
+/**
+ * Internal dependencies
+ */
+export { proxifyState, peek } from './state';
+export { proxifyStore } from './store';
```
Not sure about the naming here. I feel "proxify" is comprehensible, but maybe the appropriate word is "proxy", used as a verb.

What do you think? 🤔

luisherranz replied:
I think proxify is fine 🙂

michalczaplinski replied:
I also think `proxify` sounds all right 🙂

DAreRodz commented on packages/interactivity/src/init.ts:
```diff
@@ -29,6 +30,10 @@ export const initialVdom = new WeakMap< Element, ComponentChild[] >();
 
 // Initialize the router with the initial DOM.
 export const init = async () => {
+	// Parse and populate the initial state and config.
+	const data = parseInitialData();
+	populateInitialData( data );
+
```
I had to move the state initialization here to prevent issues with circular references between modules. ☝️

luisherranz replied:
Is that a real issue or just a Webpack warning?

DAreRodz replied:
It was a real issue. IIRC, the `proxies/state.ts` module was `undefined` while trying to import `proxifyState` inside `store.ts`. The circular dependency was like this:

```
proxies/state > hooks > store > proxies/state
```

I fixed it  in https://github.com/WordPress/gutenberg/pull/62734/commits/e1694a6da4c1b43abe0cbe9f48ea26428b5068a1 by moving the scope and namespace-related logic to their respective modules.

DAreRodz commented on packages/interactivity/src/utils.ts:
```diff
@@ -145,18 +145,26 @@ export function withScope( func: ( ...args: unknown[] ) => unknown ) {
 				} finally {
-					resetNamespace();
 					resetScope();
+					resetNamespace();
 				}
+
 				try {
 					value = await it.value;
 				} catch ( e ) {
+					setNamespace( ns );
+					setScope( scope );
 					gen.throw( e );
+				} finally {
+					resetScope();
+					resetNamespace();
 				}
+
 				if ( it.done ) {
 					break;
 				}
 			}
+
```
I introduced here the changes from https://github.com/WordPress/gutenberg/pull/62555.

DAreRodz commented:
I've marked the PR as ready for review.

DAreRodz commented on packages/interactivity/src/store.ts:
```diff
@@ -23,7 +12,10 @@ const deepMerge = ( target: any, source: any ) => {
-				Object.defineProperty( target, key, { get: getter } );
+				Object.defineProperty( target, key, {
+					get: getter,
+					configurable: true,
```
This is required in case someone wants to overwrite a getter with a value or update the getter.

luisherranz replied:
Interesting. Do we have a use case that justifies this possibility?

luisherranz replied:
The use case is when there are multiple calls to the `store` function for the same namespace.

DAreRodz commented:
@sirreal has faced an issue in this PR with the initial state merge. After taking a look, I've found these two problems we should address:

1. I moved the state initialization inside the `init()` function to avoid circular dependency problems ([here](https://github.com/WordPress/gutenberg/pull/62734/files#r1689895609)). When that code is reached, blocks have already registered their stores, so there are getters inside the state, and now `deepMerge()` reads them from `target` without doing any checks. This was not happening before, so we hadn't noticed the issue. To fix this, we could move the code to its original place and solve the circular dependency issue instead.
2. I've changed how getters are defined so now they can be overwritten ([here](https://github.com/WordPress/gutenberg/pull/62734/files#r1691315133)). This is going to fail with the current `navigate` implementation, which makes the state from the server take precedence, overwriting getters with the values from the initial state. I haven't tried it yet, but I'm sure this happens. Not sure how this part should work, though. Maybe I can revert this change until we fix the [merge problems during `navigate`](https://github.com/WordPress/gutenberg/issues/63864).

cc: @luisherranz

luisherranz commented on packages/interactivity/src/proxies/signals.ts:
```diff
@@ -0,0 +1,142 @@
+	batch,
+	type Signal,
+	type ReadonlySignal,
+} from '@preact/signals-core';
```
Could we import everything from `@preact/signals` to avoid mixing packages?

```suggestion
} from '@preact/signals';
```

DAreRodz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/62734/commits/9fb783cfe5fca6ebb3ab2b939253a90f34eb4bff.

luisherranz commented on packages/interactivity/package.json:
```diff
@@ -28,7 +28,7 @@
 	"dependencies": {
 		"@preact/signals": "^1.2.2",
-		"deepsignal": "^1.4.0",
+		"@preact/signals-core": "^1.4.0",
```
Same question here.

DAreRodz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/62734/commits/9fb783cfe5fca6ebb3ab2b939253a90f34eb4bff.

luisherranz commented on packages/interactivity/src/proxies/test/state-proxy.ts:
```diff
@@ -0,0 +1,1187 @@
+/**
+ * External dependencies
+ */
+import { effect } from '@preact/signals-core';
```
And here.

DAreRodz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/62734/commits/9fb783cfe5fca6ebb3ab2b939253a90f34eb4bff.

luisherranz commented on packages/interactivity/src/store.ts:
```diff
@@ -280,13 +148,12 @@ export function store(
-		const proxiedStore = new Proxy( rawStore, handlers );
+		const proxiedStore = proxifyStore( namespace, rawStore );
 		rawStores.set( namespace, rawStore );
 		stores.set( namespace, proxiedStore );
```
`proxyfiedStore`? 🙂

```suggestion
		const proxifiedStore = proxifyStore( namespace, rawStore );
		rawStores.set( namespace, rawStore );
		stores.set( namespace, proxifiedStore );
```

luisherranz commented:
1. Two things:

- Why don't we fix the fact that `deepMerge` calls the getters? That feels wrong to start with.
- We should make a final decision on how stores are merged.
  My current take is that merging should work this way:
  - Initial server state
  - Call `#1` to `store` - replaces existing values
  - Call `#2` to `store` - replaces existing values
  - Call `#X…` to `store` - replaces existing values
  - New server state - doesn't replace existing values
  
  If that's the final decision, then the first thing that needs to happen is the initial server state.
2. If we decide that subsequent calls (`#2`, `#X…`) to `store` should replace existing values, then we need to make the getters replaceable. So let's revert the server merge before merging this PR and that should solve the issue, shouldn't it?

luisherranz commented:
Let's also see if we can get rid of the dependency on `DOMContentLoaded`, which has been causing some issues.

luisherranz commented on packages/interactivity/src/store.ts:
```diff
@@ -1,20 +1,9 @@
+import { getNamespace } from './hooks';
+
 const isObject = ( item: unknown ): item is Record< string, unknown > =>
 	Boolean( item && typeof item === 'object' && item.constructor === Object );
```
What is the difference between this function and the one called `isPlainObject` in the `utils` file?

DAreRodz replied:
No difference at all. 😄 Fixed in https://github.com/WordPress/gutenberg/pull/62734/commits/2f376d9f300bc2ed119209324117cf470309ffa1.

luisherranz commented on packages/interactivity/src/hooks.tsx:
```diff
@@ -139,8 +139,15 @@ const namespaceStack: string[] = [];
+		throw Error(
+			'Cannot call `getContext()` outside getters and actions used by directives.'
+		);
+	}
```
It’s not just getters and actions, so we should find a more generalistic way to say this and add possible solution instructions. Apart from that, can we show the error only in development? 

Something like this (I'm not sure about the correct syntax):

```suggestion
	if ( globalThis.SCRIPT_DEBUG && ! scope ) {
		throw Error(
			'Cannot call `getContext()` when there is no scope. If you are using an async function, please consider using a generator instead. If you are using some sort of async callbacks, like `setTimeout`, please wrap the callback with `withScope(callback)`.'
		);
	}
```

We should do the same in the `getElement` function.

To be honest, I'm not sure when we should throw errors in production and when we shouldn't. I think we haven't been very consistent in that regard. Perhaps we should do a general review of all these errors to have a more consistent policy.

DAreRodz replied:
I'll include your suggestion. We can do the general review later. 🙂

luisherranz replied:
Thanks, I have included it in the [Tracking Issue](https://github.com/WordPress/gutenberg/issues/63232):  "Review all errors and make them more consistent"

luisherranz committed:
Delete unnecessary @ts-ignore-next-line

luisherranz commented on packages/interactivity/src/directives.tsx:
```diff
@@ -280,9 +286,10 @@ export default () => {
-					} );
+					updateContext(
+						currentValue.current[ namespace ],
+						deepClone( value ) as object
```
Do we need to keep doing a deep clone of the object? If I'm not mistaken, objects are no longer merged on each context, so there is only a unique object for each context. Is that correct?

DAreRodz replied:
Uhm, if I remember correctly, I added that because the parsed JSON defined in the `data-wp-context` directive―i.e., the object returned by `toVdom()` as a prop―was being modified and that caused some sort of trouble during client-side navigation.

Maybe that won't be needed once we fix https://github.com/WordPress/gutenberg/issues/63864. 🤔

luisherranz replied:
> the object returned by toVdom() as a prop―was being modified

Hmm... do you know by who? `toVdom` creates new objects on each run, doesn't it? https://github.com/WordPress/gutenberg/blob/644264cbcee25b34dd00c6ee8ba65097462e0ef0/packages/interactivity/src/vdom.ts#L105

luisherranz commented on packages/interactivity/src/proxies/registry.ts:
```diff
@@ -0,0 +1,81 @@
+	return objToProxy.get( obj ) as T;
+};
+
+/**
+ * Returns the proxy for the given object. If there is no associated proxy, the
+ * function returns `undefined`.
+ *
+ * @param obj Object from which to know the proxy.
+ * @return Associated proxy or `undefined`.
+ */
+export const getProxy = < T extends object >( obj: T ): T =>
+	objToProxy.get( obj ) as T;
+
+/**
+ * Gets the namespace associated with the given proxy.
+ *
+ * Proxies have a namespace assigned upon creation. See {@link createProxy}.
+ *
+ * @param proxy Proxy.
+ * @return Namespace.
+ */
+export const getProxyNs = ( proxy: object ): string => proxyToNs.get( proxy )!;
```
These two are not very consistent; maybe they should be called `getProxyFromObject` and `getNamespaceFromProxy` in case we need more of these in the future?

DAreRodz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/62734/commits/af00c39f536bc58975c439c853c790e610ed845c.

luisherranz commented on packages/interactivity/src/proxies/state.ts:
```diff
@@ -0,0 +1,238 @@
+		const desc = Object.getOwnPropertyDescriptor( target, key );
+		const prop = getPropSignal( receiver, key, desc );
+
+		if ( peeking ) {
+			return prop.getComputed().peek();
+		}
```
How do we get here if we have a previous conditional with `peeking` as well?

DAreRodz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/62734/commits/e8718e9145ba5ea105a9e1d49f565a0504abc38c.

luisherranz commented on packages/interactivity/src/proxies/state.ts:
```diff
@@ -0,0 +1,238 @@
+			const { get, value } = desc;
+			if ( get ) {
+				prop.setGetter( desc.get! );
+			} else {
```
Does that work?

```suggestion
				prop.setGetter( get );
```

DAreRodz replied:
I guess you mean the line above the `else` right? Yup, it works fine.

DAreRodz replied:
Updated in https://github.com/WordPress/gutenberg/pull/62734/commits/7a3cf86076d6052b4cf416789af6f26d453cdb7f.

luisherranz replied:
Yes, sorry.

luisherranz commented on packages/interactivity/src/proxies/state.ts:
```diff
@@ -0,0 +1,238 @@
+				objToIterable.get( target )!.value++;
+			}
+
+			if ( Array.isArray( target ) ) {
+				const length = getPropSignal( receiver, 'length' );
+				length.setValue( target.length );
+			}
```
We don't need to update to `length` if no one is subscribed.

Something like this?

```suggestion
			if ( Array.isArray( target ) && getPropSignal( receiver, 'length' ) ) {
				getPropSignal( receiver, 'length' ).setValue( target.length );
			}
```

DAreRodz replied:
Well, `getPropSignal` automatically creates a `PropSignal` instance, but I haven't implemented a way to know whether that `PropSignal` has been accessed. 😅 

I can think of two options:

1. Implement a `hasPropSignal` function in `proxies/state.ts` that tells whether a `PropSignal` exists for the given property. It should be as simple as this:
   ```js
   const hasPropSignal = ( proxy, key ) => proxyToProps.get( proxy )?.has( key );
   ```
2. Implement a `hasBeenAccessed` flag inside `PropSignal` instances that become true once `getComputed()` is called.

---

PS: Note that a `PropSignal`'s internal signals are not instantiated until either `getComputed()`, `setValue()`, or `setGetter()` are called.

DAreRodz replied:
I used the first option (doing the check directly in the conditional). Fixed in https://github.com/WordPress/gutenberg/pull/62734/commits/3eea77152f4cb4163c5b5502f0ae525eea697c70.

luisherranz replied:
Perfect, thank you.

luisherranz commented on packages/interactivity/src/proxies/state.ts:
```diff
@@ -0,0 +1,238 @@
+		if ( ! objToIterable.has( target ) ) {
+			objToIterable.set( target, signal( 0 ) );
+		}
+		( objToIterable as any )._ = objToIterable.get( target )!.value;
```
Let's add a comment, just in case someone wants to delete this line. Also, we should check in the minified bundle that it is indeed being preserved.

```suggestion
		// This subscribes to the signal while preventing the minifier from deleting this line in production. 
		( objToIterable as any )._ = objToIterable.get( target )!.value;
```

luisherranz commented on packages/interactivity/src/proxies/signals.ts:
```diff
@@ -0,0 +1,142 @@
+ * Structure that manages reactivity for a property in a state object. It uses
+ * signals to keep track of property value or getter modifications.
+ */
+export class PropSignal {
```
I wonder if we should change this name. Have you thought of any other options?

DAreRodz replied:
Naming is tricky. 😅

I thought of other options, but this one was the most appropriate (in my opinion). I'm open to suggestions, though! 😄

luisherranz replied:
It's internal, so let's leave it like this for now. We can change it in the future if we come up with something better 🙂

luisherranz commented on packages/interactivity/src/proxies/signals.ts:
```diff
@@ -0,0 +1,142 @@
+			this.getterSignal = signal( get );
+		} else if (
+			value !== this.valueSignal.peek() ||
+			get !== this.getterSignal?.peek()
```
Better? I suppose that `getterSignal` will always exist if we have reached this `if`, right?

```suggestion
			get !== this.getterSignal!.peek()
```

DAreRodz replied:
Updated in https://github.com/WordPress/gutenberg/pull/62734/commits/91ae5e81b16a59558e757ae4c16e6f2bde69eb12.

luisherranz reviewed (commented):
I haven't looked at the tests yet, but the code seems phenomenal to me. Very clear, very well organized, and easy to understand. Great job! 👏

I've left some comments and questions here.

luisherranz commented on packages/interactivity/src/proxies/test/store-proxy.ts:
```diff
@@ -0,0 +1,125 @@
+	getContext,
+} from '../../hooks';
+
+describe( 'interactivity api - store proxy', () => {
+	describe( 'get', () => {
```
Maybe like this?

```suggestion
describe( 'interactivity api', () => {
	describe( 'store proxy', () => {
```

DAreRodz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/62734/commits/75d088908bd5e8782df88b61215151ea1e43de4f.

luisherranz commented on packages/interactivity/src/proxies/test/state-proxy.ts:
```diff
@@ -0,0 +1,1187 @@
+	}
+};
+
+describe( 'interactivity api - state proxy', () => {
```
Same here?


```suggestion
describe( 'interactivity api', () => {
	describe( 'state proxy', () => {
```

And we should also change it in the tests of the `utils` file.

DAreRodz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/62734/commits/75d088908bd5e8782df88b61215151ea1e43de4f.

luisherranz commented on packages/interactivity/src/proxies/test/state-proxy.ts:
```diff
@@ -0,0 +1,1187 @@
+			} );
+
+			const scope = { context: { test: { value: 'from context' } } };
+			try {
+				setScope( scope as any );
+				setNamespace( 'test' );
+				expect( state.y ).toBe( 'from context' );
+			} finally {
+				resetNamespace();
+				resetScope();
+			}
```
Shouldn't namespace be interally set?

The test pass with this:

```suggestion
			try {
				setScope( scope as any );
				expect( state.y ).toBe( 'from context' );
			} finally {
				resetScope();
			}
```

DAreRodz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/62734/commits/8a6589aba4a69b7f339a327ffa3543f2b3ea8ef0

luisherranz commented on packages/interactivity/src/proxies/test/state-proxy.ts:
```diff
@@ -0,0 +1,1187 @@
+			} );
+
+			const scope = { context: { test: { value: 1 } } };
+			try {
+				setScope( scope as any );
+				setNamespace( 'test' );
+				expect( state.sumContextValue( 2 ) ).toBe( 3 );
+			} finally {
+				resetNamespace();
+				resetScope();
+			}
```
Same here.

```suggestion
			try {
				setScope( scope as any );
				expect( state.sumContextValue( 2 ) ).toBe( 3 );
			} finally {
				resetScope();
			}
```

DAreRodz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/62734/commits/8a6589aba4a69b7f339a327ffa3543f2b3ea8ef0

luisherranz commented on packages/interactivity/src/proxies/test/state-proxy.ts:
```diff
@@ -0,0 +1,1187 @@
+			expect( state.array.length ).toBe( 1 );
+		} );
+
+		it( 'should update when mutations happen', () => {
+			expect( state.a ).toBe( 1 );
+			state.a = 11;
+			expect( state.a ).toBe( 11 );
+		} );
```
What is the difference between this and 'should update like plain objects/arrays'?

DAreRodz replied:
They look the same. I'll remove this one.

BTW, those tests came from `luisherranz/deepsignal` (see [this](https://github.com/luisherranz/deepsignal/blob/8fb50d677615b368eae90a4e8b9e4570fa459f62/packages/deepsignal/core/test/index.test.tsx#L183-L190) and [this](https://github.com/luisherranz/deepsignal/blob/8fb50d677615b368eae90a4e8b9e4570fa459f62/packages/deepsignal/core/test/index.test.tsx#L223-L227)), so maybe it should be fixed there as well? 🤔

DAreRodz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/62734/commits/c7f13ad847fd634b7c13e72b04a594f2e975f8a4.

luisherranz replied:
> BTW, those tests came from `luisherranz/deepsignal` (see [this](https://github.com/luisherranz/deepsignal/blob/8fb50d677615b368eae90a4e8b9e4570fa459f62/packages/deepsignal/core/test/index.test.tsx#L183-L190) and [this](https://github.com/luisherranz/deepsignal/blob/8fb50d677615b368eae90a4e8b9e4570fa459f62/packages/deepsignal/core/test/index.test.tsx#L223-L227)), so maybe it should be fixed there as well? 🤔

I'll fix that, thank you!

luisherranz commented on packages/interactivity/src/proxies/test/state-proxy.ts:
```diff
@@ -0,0 +1,1187 @@
+			expect( state.a ).toBe( 11 );
+		} );
+
+		it( 'should support setting getters on the fly', () => {
```
Maybe we should add a test similar to this one, but where `state.double` does exist (it's not `undefined`).

DAreRodz replied:
Done in https://github.com/WordPress/gutenberg/pull/62734/commits/378bc766f6ad83b7faf52a2e3aa89c035abbe43e. I've also added another test to check that the effects are re-executed.

luisherranz commented on packages/interactivity/src/proxies/test/state-proxy.ts:
```diff
@@ -0,0 +1,1187 @@
+			expect( raw2.obj ).toBe( state1.obj );
+		} );
+
+		it( 'should use its namespace by default inside setters', () => {
```
Why specifically within setters (and not within getters, for example)?

DAreRodz replied:
I've added a test case for that in https://github.com/WordPress/gutenberg/pull/62734/commits/8a00d34d945fa32342a808f44da538af2ee7e0a2.

luisherranz commented on packages/interactivity/src/proxies/test/state-proxy.ts:
```diff
@@ -0,0 +1,1187 @@
+			expect( state1 ).toBe( state2 );
+		} );
+
+		it( 'should return the same proxy when trying to re-proxify a state object', () => {
```
This?

```suggestion
		it( 'should throw when trying to re-proxify a state object', () => {
```

DAreRodz replied:
Nice catch. 😄 Fixed in https://github.com/WordPress/gutenberg/pull/62734/commits/417a0c591c2ce0b57bbc0fc83fa2c5529404d9fa.

luisherranz reviewed (commented):
I took a look at the tests, very good job! 

I have left other comments.

gziolo commented:
Nice to see a slight decrease in the bundle size:

![image](https://github.com/user-attachments/assets/8706247d-b5c7-4d4a-b31a-dd7b7a555803)

DAreRodz committed:
Use `isPlainObject` from utils inside store

DAreRodz committed:
Move scopes and namespaces to separate files

DAreRodz committed:
Call `init` outside `DOMContentLoaded`

DAreRodz committed:
Replace `signals-core` imports with `signals`

DAreRodz committed:
Rename `proxiedStore` to `proxifiedStore`

Co-authored-by: Luis Herranz <luisherranz@gmail.com>

DAreRodz committed:
Remove unnecessary `peek()` call

DAreRodz committed:
Throw more descriptive errors from getContext and getElement

DAreRodz committed:
Use more descriptive name for proxy functions

DAreRodz committed:
Use destructured `get` inside `setGetter` call

DAreRodz committed:
Add comment to explain `objToIterable` signals subscription

Co-authored-by: Luis Herranz <luisherranz@gmail.com>

DAreRodz committed:
Change line comment to block comment

DAreRodz committed:
Replace `?` with `!` operator

DAreRodz committed:
Wrap Interactivity API tests with appropriate `describe`

DAreRodz committed:
Remove unnecessary `setNamespace` calls in tests

DAreRodz committed:
Remove duplicated test

DAreRodz committed:
Fix typo

DAreRodz committed:
Add extra tests for getter modification

DAreRodz committed:
Test the right namespace is used inside getters

DAreRodz committed:
Fix test name

DAreRodz committed:
Check if length's PropSignal exists before updating its value

DAreRodz commented:
Luis, I've addressed all the actionable feedback you shared. Feel free to take a look at the PR again. 🙂

luisherranz committed:
Add deepMerge tests and prevent server overwritting

luisherranz committed:
Make sure context inheritance is shallow and server props don't overwrite

luisherranz committed:
Refactor wp-each to use new proxifyContext structure

luisherranz reviewed (approved):
Let's merge this so it can be included in the next release.

I'll add more tests in a new PR to try to ensure we haven't missed anything.
